ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,372 @@
+import qualified Ouroboros.Consensus.Storage.LedgerDB.HD as HD
+import qualified Ouroboros.Consensus.Ledger.Basics as Tables
+import qualified Control.Exception as Exn
+import qualified Ouroboros.Consensus.Storage.FS.API as FS
+import Cardano.Slotting.Slot (SlotNo, WithOrigin (At))
+import Ouroboros.Consensus.Util.IOLike (IOLike)
+import qualified Control.Monad.Class.MonadAsync as IOLike
+import qualified Control.Monad.Class.MonadSTM as IOLike
+import Control.Monad.IO.Class (MonadIO(liftIO))
+import Control.Monad ( forever, unless, when, void )
+import Data.Functor ( ($>), (<&>) )
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+import Codec.Serialise (Serialise)
+import Data.Map (Map)
+import qualified Data.Set as Set
+import Ouroboros.Consensus.Util (foldlM')
+import Control.Monad.Trans.Maybe
+import Control.Monad.Trans.Class
+import qualified Database.LMDB.Simple.Extra as LMDB
+import Data.Foldable (for_)
+import Database.LMDB.Simple (ReadWrite)
+import qualified Ouroboros.Consensus.Storage.FS.API.Types as FS
+
+{-
+
+TODO Don't use DbRef
+
+-}
+{-# HLINT ignore #-}
+
+type LMDBBackingStore l = HD.BackingStore IO
+  (Tables.LedgerTables l Tables.KeysMK)
+  (Tables.LedgerTables l Tables.ValuesMK)
+  (Tables.LedgerTables l Tables.DiffMK)
+-- type LMDBBackingStore k v = HD.BackingStore IO (UtxoKeys k v) (UtxoValues k v) (UtxoDiff k v)
+
+type LMDBValueHandle l = HD.BackingStoreValueHandle IO
+  (Tables.LedgerTables l Tables.KeysMK)
+  (Tables.LedgerTables l Tables.ValuesMK)
+  
+
+
+data Db l = Db
+  { dbEnv :: !(LMDB.Environment LMDB.ReadWrite)
+  , dbSettings :: !(LMDB.DBRef LMDB.ReadWrite DbState)
+  , dbValueHandles :: !(IOLike.TMVar IO (Map (WithOrigin SlotNo) ValueHandle))
+  , dbBackingTables :: !(Tables.LedgerTables l Tables.LMDBMK)
+  , dbFilePath :: !FilePath
+  }
+
+data DbErr = DbErrStr !String
+  | DbErrNoSettings
+  | DbErrNonMonotonicSeq !(WithOrigin SlotNo) !(WithOrigin SlotNo)
+  | DbErrNoDbNamed !String
+  | DbErrBadDynamic !String
+  | DbErrInitialisingNonEmpty !String
+  | DbErrBadRead
+  deriving stock (Show)
+
+instance Exn.Exception DbErr
+
+data DbState  = DbState
+  { dbsSeq :: !(WithOrigin SlotNo)
+  -- TODO a version field
+  } deriving (Show, Generic)
+
+instance Serialise (DbState)
+
+readLMDBTable :: (Ord k, Serialise k, Serialise v)
+  => LMDB.Database k v
+  -> HD.UtxoKeys k v
+  -> LMDB.Transaction ReadWrite (HD.UtxoValues k v)
+readLMDBTable db (HD.UtxoKeys keys) =
+  HD.UtxoValues <$> foldlM' go Map.empty (Set.toList keys) where
+  go m k = LMDB.get db k <&> \case
+    Nothing -> m
+    Just v -> Map.insert k v m
+
+writeLMDBTable :: (Serialise k, Serialise v)
+  => LMDB.Database k v
+  -> HD.UtxoDiff k v
+  -> LMDB.Transaction LMDB.ReadWrite ()
+writeLMDBTable db (HD.UtxoDiff m) = void $ Map.traverseWithKey go m where
+  go k (HD.UtxoEntryDiff v reason) = LMDB.put db k new_val where
+    new_val
+      | reason `elem` [HD.UedsDel, HD.UedsInsAndDel]
```

Comment:
`UedsInsAndDel` can be a no-op; no `put` call necessary.

Reply:
For the UTxO map, if there is an existing value on disk, then no valid diff can involving inserting at that key. So those three steps as listed cannot arise from a valid chain suffix (which these must be, since they're being placed in the immutable ChainDB).

Does that logic seem sound to you?